#Mounting Google Drive and Importing Necessary Packages

In [1]:
from google.colab import drive
drive.mount('/content/drive')
%cd drive/MyDrive/EPFL/MA3/ADA/Project_Data

Mounted at /content/drive
/content/drive/MyDrive/EPFL/MA3/ADA/Project_Data


In [2]:
import numpy as np
import pandas as pd
import os

#Loading Data to Pandas Dataframes and Preliminary Analysis

##Loadings

In [3]:
path = '/wikispeedia_paths-and-graph'

In [4]:
article_df = pd.read_csv(os.getcwd() + path + '/articles.tsv', sep = '\t', header = None, comment = '#', names = ['Article Names'])
article_df

Article Names
0     %C3%81ed%C3%A1n_mac_Gabr%C3%A1in
1                           %C3%85land
2                   %C3%89douard_Manet
3                            %C3%89ire
4           %C3%93engus_I_of_the_Picts
...                                ...
4599                           Zionism
4600                         Zirconium
4601                         Zoroaster
4602                      Zuid-Gelders
4603                              Zulu

[4604 rows x 1 columns]

In [5]:
categ_df = pd.read_csv(os.getcwd() + path + '/categories.tsv', sep = '\t', header = None, comment = '#', names = ['Article Names', 'Categories'])
categ_df

Article Names  \
0     %C3%81ed%C3%A1n_mac_Gabr%C3%A1in   
1     %C3%81ed%C3%A1n_mac_Gabr%C3%A1in   
2                           %C3%85land   
3                           %C3%85land   
4                   %C3%89douard_Manet   
...                                ...   
5199                         Zirconium   
5200                         Zoroaster   
5201                      Zuid-Gelders   
5202                      Zuid-Gelders   
5203                              Zulu   

                                             Categories  
0     subject.History.British_History.British_Histor...  
1                     subject.People.Historical_figures  
2                                     subject.Countries  
3     subject.Geography.European_Geography.European_...  
4                                subject.People.Artists  
...                                                 ...  
5199        subject.Science.Chemistry.Chemical_elements  
5200       subject.People.Religious_figures_and_leaders  
5201               subject.Geography.European_Geography  
5202          subject.Language_and_literature.Languages  
5203                subject.Geography.African_Geography  

[5204 rows x 2 columns]

In [14]:
links_df = pd.read_csv(os.getcwd() + path + '/links.tsv', sep = '\t', header = None, comment = '#', names = ['Article Names', 'Links'])
links_df.head()

Article Names           Links
0  %C3%81ed%C3%A1n_mac_Gabr%C3%A1in            Bede
1  %C3%81ed%C3%A1n_mac_Gabr%C3%A1in         Columba
2  %C3%81ed%C3%A1n_mac_Gabr%C3%A1in  D%C3%A1l_Riata
3  %C3%81ed%C3%A1n_mac_Gabr%C3%A1in   Great_Britain
4  %C3%81ed%C3%A1n_mac_Gabr%C3%A1in         Ireland

In [15]:
paths_fin_df = pd.read_csv(os.getcwd() + path + '/paths_finished.tsv', sep = '\t',header = None, comment = '#', names = ['Hashed IP Address', 'Timestamp', 'Duration(s)', 'Path', 'Rating'])
paths_fin_df.head()

Hashed IP Address   Timestamp  Duration(s)  \
0  6a3701d319fc3754  1297740409          166   
1  3824310e536af032  1344753412           88   
2  415612e93584d30e  1349298640          138   
3  64dd5cd342e3780c  1265613925           37   
4  015245d773376aab  1366730828          175   

                                                Path  Rating  
0  14th_century;15th_century;16th_century;Pacific...     NaN  
1  14th_century;Europe;Africa;Atlantic_slave_trad...     3.0  
2  14th_century;Niger;Nigeria;British_Empire;Slav...     NaN  
3     14th_century;Renaissance;Ancient_Greece;Greece     NaN  
4  14th_century;Italy;Roman_Catholic_Church;HIV;R...     3.0

In [18]:
paths_unfin_df = pd.read_csv(os.getcwd() + path + '/paths_unfinished.tsv', sep = '\t', header = None, comment = '#', names = ['Hashed IP Address', 'Timestamp', 'Duration(s)', 'Path', 'Target','Type'])
paths_unfin_df.head()

Hashed IP Address   Timestamp  Duration(s)  \
0  2426091a53125110  1297054935         1804   
1  26141fd878806294  1297055651         1805   
2  2b015fb8181c48f2  1297090819         1818   
3  53a53bc244e08a6a  1297094761           49   
4  53a53bc244e08a6a  1297099105         1808   

                                   Path            Target     Type  
0                        Obi-Wan_Kenobi         Microsoft  timeout  
1                         Julius_Caesar           Caracas  timeout  
2  Malawi;Democracy;Alexander_the_Great     First_Crusade  timeout  
3                              Paraguay  Mount_St._Helens  restart  
4                      Paraguay;Bolivia  Mount_St._Helens  timeout

In [19]:
shrt_path_distance_df = pd.read_csv(os.getcwd() + path + '/shortest-path-distance-matrix.txt', sep = '\t', header = None, comment = '#', names = ['Shortest Paths'])
shrt_path_distance_df.head()

Shortest Paths
0  0_____33333325634333435_2433544334_3_422343544...
1  _0____22222325623232424_2422544324_3_312242544...
2  __0___33222425623232324_2333444433_3_422343434...
3  ___0__33333325634233334_2433434333_2_423343433...
4  ____0_22323335633332435_2433545434_3_423343544...

##Preliminary Analysis on Categories and Finished/Unfinished Paths

###Categories

In [7]:
categories = categ_df['Categories']
categories

0       subject.History.British_History.British_Histor...
1                       subject.People.Historical_figures
2                                       subject.Countries
3       subject.Geography.European_Geography.European_...
4                                  subject.People.Artists
                              ...                        
5199          subject.Science.Chemistry.Chemical_elements
5200         subject.People.Religious_figures_and_leaders
5201                 subject.Geography.European_Geography
5202            subject.Language_and_literature.Languages
5203                  subject.Geography.African_Geography
Name: Categories, Length: 5204, dtype: object

In [36]:
main_cat_list = []
sec_cat_list = []
thrd_cat_list = []
length = 0
for cat in categories:
  splitted = cat.split('.')
  if splitted[1] not in main_cat_list :
      main_cat_list.append(splitted[1])

  try:
    if splitted[2] not in sec_cat_list:
      sec_cat_list.append(splitted[2])
    if splitted[3] not in thrd_cat_list:
      thrd_cat_list.append(splitted[3])
  except:IndexError

  finally:
    pass

print(f'There are {len(main_cat_list)} unique main categories, {len(sec_cat_list)} unique second level categories, and {len(thrd_cat_list)} unique third level categories.')

There are 15 unique main categories, 101 unique second level categories, and 27 unique third level categories


####Discussion:
As we have 4604 articles with 15 main, 101 second level, 27 third level categories, we will be able to make more genera observations, if there is any, on how humans make connections between different categories.

###Finished & Unfinished Paths

In [51]:
print(f"Although there are {len(paths_fin_df)} finished paths in the data, there are {len(paths_fin_df['Hashed IP Address'].unique())} players which means there \nare several finished paths belonging to the same player which might contain important information.")

Although there are 51318 finished paths in the data, there are 14245 players which means there 
are several finished paths belonging to the same player which might contain important information 


In [57]:
grouped_fin_df = paths_fin_df.groupby('Hashed IP Address')
duplica_fin_df = grouped_fin_df.filter(lambda x: len(x) > 1)
print(f" {len(duplica_fin_df['Hashed IP Address'].unique())} players have more than one finished paths in the data.")


 7222 players have more than one finished paths in the data.


In [67]:
#example of a player that finished the game four times
paths_fin_df[paths_fin_df['Hashed IP Address'] == '6a3701d319fc3754']

Hashed IP Address   Timestamp  Duration(s)  \
0      6a3701d319fc3754  1297740409          166   
18730  6a3701d319fc3754  1297740041          452   
26263  6a3701d319fc3754  1297740209          128   
26413  6a3701d319fc3754  1297740992           37   

                                                    Path  Rating  
0      14th_century;15th_century;16th_century;Pacific...     NaN  
18730  Land_%28economics%29;Economics;John_Maynard_Ke...     NaN  
26263           Sugar;Industrial_Revolution;Frankenstein     NaN  
26413             Thunderstorm;United_States;Puerto_Rico     NaN

In [55]:
print(f"Following the same logic there are {len(paths_unfin_df)} unfinished paths and {len(paths_unfin_df['Hashed IP Address'].unique())} players in the data which means there \nare several unfinished paths belonging to the same player which might again contain important information.")

Following the same logic there are 24875 unfinished paths and 11485 players in the data which means there 
are several unfinished paths belonging to the same player which might again contain important information.


In [58]:
grouped_unfin_df = paths_unfin_df.groupby('Hashed IP Address')
duplica_unfin_df = grouped_unfin_df.filter(lambda x: len(x) > 1)
print(f" {len(duplica_unfin_df['Hashed IP Address'].unique())} players have more than one unfinished paths in the data.")


 4399 players have more than one unfinished paths in the data.


In [68]:
#example of a player that have 5 unfinished paths
paths_unfin_df[paths_unfin_df['Hashed IP Address'] == '6d136e371e42474f']


Hashed IP Address   Timestamp  Duration(s)  \
7     6d136e371e42474f  1297102070          175   
8     6d136e371e42474f  1297102100           19   
9     6d136e371e42474f  1297104110         2006   
581   6d136e371e42474f  1298673754         1866   
1585  6d136e371e42474f  1299618558         1807   

                                                   Path         Target  \
7       4-2-0;United_States;Agriculture;Sugar;Australia      Cane_Toad   
8                                                 Logic   Moon_landing   
9     Logic;Computer_science;Science;Physical_scienc...   Moon_landing   
581        Nostradamus;Black_Death;Cattle;Animal;Mammal  Animal_rights   
1585                                       Fidel_Castro  Domesday_Book   

         Type  
7     restart  
8     restart  
9     timeout  
581   timeout  
1585  timeout

In [70]:
fin_IP_set = set(paths_fin_df['Hashed IP Address'].unique())
unfin_IP_set = set(paths_unfin_df['Hashed IP Address'].unique())
players_fin_unfin = fin_IP_set.intersection(unfin_IP_set)
print(f"{len(players_fin_unfin)} players have both finished and unfinished paths which might contain important information about the success of the players.")

4862 players have both finished and unfinished paths which might contain important information about the success of the players.


In [75]:
paths_fin_df[paths_fin_df['Hashed IP Address'] == '7dabfd2e670a9e8d']

Hashed IP Address   Timestamp  Duration(s)  \
14923  7dabfd2e670a9e8d  1353189333           80   
26655  7dabfd2e670a9e8d  1353189031           18   
48898  7dabfd2e670a9e8d  1353189213          124   

                                                    Path  Rating  
14923    Brain;Sleep;United_States;Electronics;Telephone     NaN  
26655       RMS_Titanic;England;London;Buckingham_Palace     NaN  
48898  Theatre;United_States;American_Civil_War;Slave...     NaN

In [76]:
paths_unfin_df[paths_unfin_df['Hashed IP Address'] == '7dabfd2e670a9e8d']

Hashed IP Address   Timestamp  Duration(s)  \
14575  7dabfd2e670a9e8d  1353189002          131   
14576  7dabfd2e670a9e8d  1353189067           25   

                                                    Path           Target  \
14575  Zinc;German_language;Brothers_Grimm;Berlin;Wor...  Hebrew_language   
14576             History_of_South_Carolina;Mercantilism             Beer   

          Type  
14575  restart  
14576  restart